cpp/src/neighbors/ivf_pq/ivf_pq_build.cuh

```diff
@@ -1515,12 +1517,22 @@ void extend(raft::resources const& handle,
     n_rows + (kIndexGroupSize - 1) * std::min<IdxT>(n_clusters, n_rows));
 
   // Available device memory
-  size_t free_mem, total_mem;
-  RAFT_CUDA_TRY(cudaMemGetInfo(&free_mem, &total_mem));
-
+  size_t free_mem = raft::resource::get_workspace_free_bytes(handle);
+
+  // We try to use the workspace memory by default here.
+  // If the workspace limit is too small, we change the resource for batch data to the
+  // `large_workspace_resource`, which does not have the explicit allocation limit. The user may opt
+  // to populate the `large_workspace_resource` memory resource with managed memory for easier
+  // scaling.
+  rmm::device_async_resource_ref labels_mr  = device_memory;
+  rmm::device_async_resource_ref batches_mr = device_memory;
+  if (n_rows * (index->dim() * sizeof(T) + index->pq_dim() + sizeof(IdxT) + sizeof(uint32_t)) >
+      free_mem) {
+    labels_mr = large_memory;
+  }
   // Allocate a buffer for the new labels (classifying the new data)
-  rmm::device_uvector<uint32_t> new_data_labels(n_rows, stream, device_memory);
-  free_mem -= sizeof(uint32_t) * n_rows;
+  rmm::device_uvector<uint32_t> new_data_labels(n_rows, stream, labels_mr);
+  if (labels_mr == device_memory) { free_mem -= sizeof(uint32_t) * n_rows; }
```

Comment:
I believe the value returned by `get_workspace_free_bytes` does not take into account the alignment/padding. The reason is that the workspace resource is a limiting adapter around the (possibly pooled, padded, aligned) memory resource; it counts the bytes requested by the user, not the bytes allocated on the device. But this is ok: this whole construct does not represent the actual device memory anyway; the workspace resource is supposed to be limited to a fraction of the device memory, so it should allow extra padding bytes with no problem.

So I think the original and your suggested code should result in identical update to `free_mem`. Do you think I still better apply the suggestion for readability?

Reply:
It does  take [alignment into account](https://github.com/rapidsai/rmm/blob/dcca09818f1c5b5b1936f684b0ff9cca5683e63f/include/rmm/mr/device/limiting_resource_adaptor.hpp#L125-L126). But in general you are right, it does not consider all possible factors. So yes, I accept that for our purpose it would be fine to track size in `free_mem` variable manually. 

I think the suggested code is simpler to read, please apply the suggestions.